### PR TITLE
feat(space): read-only list_subscriptions diagnostic tool [#908]

### PR DESCRIPTION
**Pull request by lsm**
PR 6 of the external-event subscription refactor (depends on #904 / PR 5). Adds a read-only `list_subscriptions` MCP tool that snapshots a run's external-event subscriptions across three layers — **declared** (workflow definition), **persisted** (the PR 5 table), and **active** (in-memory trie, cross-check only) — so an agent can answer "is this node actually subscribed to these events?" from durable state alone. The durable layers are the source of truth; the trie is shown only as a sanity check, with each active entry reconciled against durable state (`source: declared|persisted|orphan`) and durable rows missing from the trie surfaced via `active: false` plus a `mismatches` summary. For task #896 this would have shown "Coding node has no declared PR-event interest" from durable data alone.

Additive plumbing: `TopicTrie.values()` (diagnostic enumeration), `SpaceWorkflowEventSubscriptionRepository.listByRun()` (run-scoped read — PR 5 removed it as unused; now needed), `SpaceRuntime.listSubscriptions()` (three-layer snapshot + reconciliation, run/node scoped, space guarded), the `list_subscriptions` node-agent tool, and the `SpaceRuntimeService` facade + `TaskAgentManager` wiring. Unit tests cover all three layers, the declared/persisted/orphan reconciliation, the nodeId filter, the cross-space guard, and the #896 scenario.

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** HyperNeo | **Provider:** Anthropic

**Recommendation: REQUEST_CHANGES** — 0 P0, 1 P1, 4 P2, 5 P3. Approvable after the P1 (terminal-task false-positive `declaredNotActive` on completed runs) is addressed; the rest are polish.

### Summary

PR #6 of the external-event subscription refactor adds a read-only `list_subscriptions` node-agent diagnostic that snapshots a run's subscriptions across three layers — **declared** (workflow definition, durable), **persisted** (the PR-5 dynamic table, durable), **active** (in-memory trie, cross-check only) — and reconciles them via normalized `slot + topic` keys. This is a strong, well-scoped addition: layering is clean (tool → callback → facade → runtime → repo), the `spaceId` cross-space guard is sound (sourced from the trusted session `task.spaceId`, never an agent arg), SQL is parameterized, the path is provably read-only, and the durable-model layering correctly matches PR-5 (declared = re-derived from definition; persisted = dynamic rows only). Tests assert real computed values (specific topics, `active`/`source` flags, exact mismatch counts) rather than smoke-checking shapes, and the #896 scenario test genuinely proves the diagnostic answers the original question.

`getWorkflowForRun` resolves through the run's **pinned** definition version (RFC §4), so the `declared` layer reflects exactly what the run executes — a subtle but important consistency point with how static interests are materialized. `TopicTrie.values()` correctly walks disjoint `exact`/`glob` children with no duplication.

### Findings

**P1 — `space-runtime.ts:1414–1441` — Terminal runs report false-positive `declaredNotActive` drift.** `listSubscriptions` unconditionally re-derives declared interests from the workflow definition, but `registerRunInterestsFromWorkflow` (line 1186) deliberately *skips* static re-materialization when the canonical task is `done`/`archived`/`cancelled` (its static interests are cleared by the task lifecycle). So a diagnostic against a completed run shows every declared interest as `active: false` and counts them all in `mismatches.declaredNotActive` — looking exactly like drift when it is intentional lifecycle cleanup. For a diagnostic whose stated job is surfacing declared-vs-active discrepancies, this is the worst case: the headline signal is wrong for the common "investigate a finished run" use case. Either (a) gate `declaredNotActive` on the canonical task being non-terminal (mirror line 1186), or (b) document in the result/doc that `active:false` is expected for terminal tasks and exclude terminal-task runs from the count. Prefer (a) so the mismatch count only fires for runs where a static entry *should* be live.

**P2 — `space-runtime.ts:433` (doc) — `source: 'orphan'` framing misleads on definition edits.** The doc calls orphan static entries "drift that should not happen," but they also legitimately arise when a workflow definition is edited *after* the run started (the trie still holds the now-removed static entry until next rehydrate). That is benign, not an invariant violation. Narrow the doc to note the upgrade-window semantics, or the operator misreads a normal edit as a bug. (The data itself is correct.)

**P2 — `space-runtime.ts:1458–1463` — O(total-subscriptions-in-space) per call.** `topicTrie.values()` walks the *entire* per-space trie (including long-horizon-agent entries, filtered out after) then `.filter()`s by `workflowRunId` — twice. For a space with N concurrent runs × M subscriptions, every call is O(N·M) plus a full array allocation, whereas the sibling `list_deliveries` is an indexed, `LIMIT`-bounded SQL query. Acceptable for an occasionally-invoked diagnostic at current scale, but worth a `valuesByRun(workflowRunId)` trie method or a secondary `Map<runId, targets[]>` index if this is ever called in a loop or spaces grow large. Not blocking.

**P2 — `node-agent-tools.test.ts:3088–3163` — No end-to-end pass-through test of the result envelope.** The wiring tests assert the callback receives exact forwarded args (good), but none wire `onListSubscriptions` to a mock returning a realistic `{success:true, result:{declared:[…], mismatches:{…}}}` and assert it survives the `jsonResult` round-trip. The runtime-direct tests cover the structure and these cover the wiring; the gap is narrow but a single richer-payload assertion closes it.

**P2 — `space-runtime.ts:403–460` — Four interfaces `export`ed with zero external consumers.** `SubscriptionDeclaredInterest`/`SubscriptionPersistedRow`/`SubscriptionActiveEntry`/`SubscriptionListResult` are never imported outside this file (the facade uses `ReturnType<SpaceRuntime['listSubscriptions']>`). The sibling `WorkflowSubscriptionTarget` is kept internal; drop the `export` to match and avoid a premature public surface.

**P3 — `space-runtime-list-subscriptions.test.ts` — Missing edge-case coverage.** No test for: (1) multiple static interests on the same slot (confirms no `Set`-collapse across topics), (2) topic casing differences between declared and the materialized trie entry, (3) a run whose workflow was deleted (`getWorkflowForRun → null`, `declared: []`, persisted/active still reconcile without crashing), (4) idempotent `registerRunInterests` re-registration not double-counting in `active`. The implementation handles all four; a test each would lock them in.

**P3 — `node-agent-tools.test.ts:3143–3163` — `_registeredTools` introspection is implementation-coupled.** Reaching into `instance._registeredTools` via a double cast is brittle to SDK bumps. Consistent with the existing pattern in this file, so acceptable, but flag for awareness.

**P3 — `node-agent-tools.ts:2155` vs `node-agent-tool-schemas.ts:213` — minor vocab drift.** The schema docstring says "static interests" where the MCP tool description omits "static." Align the wording so an agent reading the schema vs the listing sees the same vocabulary.

**P3 — `node-agent-tools.ts:9–15` — Header "Discovery tools (read-only)" list omits the event tools.** Predates this PR (`list_deliveries`, `subscribe_external_event`, etc. already absent), but this PR adds another unlisted entry. Update or trim the comment to representative examples.

### Verified clean

Authorization & cross-space guard, parameterized SQL (no injection), read-only purity (`values()`/`listByRun`/reconciliation are pure reads), reconciliation key casing alignment across all three layers, `topicFrom` handling (inert, excluded from mismatch count — consistent with never entering the trie), the P1-recovery dedup/backfill in `registerSubscription`, the discriminated-union return type resolving through the facade, conditional MCP registration gated on `onListSubscriptions`, and the PR-5 `listByRun` re-add (confirmed absent at `cef74b9d5`). No P0. No security findings. CI is green.

*(Own-PR: GitHub rejects APPROVE/REQUEST_CHANGES from the author, so this is posted as a COMMENT review carrying `Recommendation: REQUEST_CHANGES`.)*

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 1]
```diff
@@ -388,6 +388,82 @@ function isLongHorizonSubscriptionTarget(
...
+    /** Persisted rows with no matching active dynamic entry. */
+    persistedNotActive: number;
+    /** Active entries with no durable backing (static w/o declared, dynamic w/o persisted). */
+    orphanActive: number;
+  };
+}
+
+// Normalized reconciliation key: slot (node + agent) + lowercased topic.
+function subscriptionReconcileKey(nodeId: string, agentName: string, topic: string): string {
+  return `${nodeId}|${agentName.toLowerCase()}|${topic.toLowerCase()}`;
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts]: Fixed in `a6a6b839e`: the dynamic reconcile key now includes `taskId` (static keys stay task-independent since declared interests are slot-level). Two tasks sharing a slot+agent+topic no longer cross-match — a trie entry for task A won't mark task B's persisted row active. Test added (`dynamic reconciliation is scoped by taskId …`).

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 2]
```diff
@@ -1304,6 +1380,148 @@ export class SpaceRuntime {
...
+    return {
+      success: true,
+      result: {
+        workflowRunId,
+        nodeId: nodeFilter,
+        declared,
+        persisted,
+        active,
+        mismatches: {
+          declaredNotActive: declared.filter((d) => d.topic !== null && !d.active).length,
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts]: Fixed in `774711087`: `declaredNotActive` is now gated on the canonical task being non-terminal, mirroring `registerRunInterestsFromWorkflow`'s skip gate, so a finished/cancelled run's cleared static interests read as expected lifecycle cleanup rather than drift. Test added.

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 3]
```diff
@@ -1304,6 +1380,148 @@ export class SpaceRuntime {
...
+    }
+    if (run.spaceId !== spaceId) {
+      return { success: false, error: `Workflow run ${workflowRunId} is not in this space.` };
+    }
+    const workflow = this.config.spaceWorkflowManager.getWorkflowForRun(run) ?? null;
+    const nodeFilter = nodeId ?? null;
+
+    // Layer 1 — declared static interests, re-derived from the definition.
+    const declared: SubscriptionDeclaredInterest[] = [];
+    if (workflow) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts:1462]: Addressed in `a6a6b839e`: the result now carries a `definitionResolved` flag. When the run's workflow definition can't be loaded, `declared` is empty and `definitionResolved:false` distinguishes the unavailable durable layer from a valid workflow with no interests — so it isn't misread as "node declares no subscriptions." Persisted/active still reconcile. Test updated.

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 4]
```diff
@@ -1304,6 +1386,177 @@ export class SpaceRuntime {
...
+        nodeId: nodeFilter,
+        declared,
+        persisted,
+        active,
+        mismatches: {
+          // Only count as drift when a static entry should be live; for terminal
+          // tasks static interests are intentionally cleared (see above).
+          declaredNotActive: staticMaterializable
+            ? declared.filter((d) => d.topic !== null && !d.active).length
+            : 0,
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts:1606]: Fixed in `1ecbb2b61`: a static entry is now "backed" only while the canonical task is non-terminal. On a terminal task, any surviving static entry is reclassified `source:'orphan'` and counted in `orphanActive`, closing the inverse blind spot (stale cleanup hidden behind the `declaredNotActive` suppression). Test added.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** HyperNeo | **Provider:** Anthropic

**Recommendation: APPROVE** — 0 P0, 0 P1, 0 P2, 2 P3. All prior findings addressed; all four inline threads resolved. Clean to merge.

### Re-review summary

Re-inspected at `1ecbb2b61e` (head stable since inspection; all substantive CI green — only the routine Coverage Quality Gate pending and conditional jobs skipping). Verified that each previously-raised finding is resolved in the current diff and traced the surrounding code fresh.

**Prior P1 — terminal-run false-positive `declaredNotActive` (space-runtime.ts).** Fixed correctly. `listSubscriptions` now computes `staticMaterializable = !!canonicalTask && status not in {done, archived, cancelled}`, which exactly mirrors `registerRunInterestsFromWorkflow`'s skip gate (line 1208) and early `if (!task) return` (line 1198). `declaredNotActive` is gated on it, so a finished/cancelled run's cleared static interests read as expected lifecycle cleanup, not drift. The inverse blind spot Codex flagged is also closed: a stale static trie entry on a terminal task is reclassified `source: 'orphan'` and counted in `orphanActive` (the `backed` check requires `staticMaterializable`), so the suppression can no longer hide leftover cleanup. Both directions are locked in by tests (line 386 suppression, line 418 orphan inverse) asserting exact mismatch counts.

**Prior P2 — dynamic reconcile key cross-match (Codex).** Fixed. `subscriptionReconcileKey` includes `taskId` for dynamic rows/targets and omits it for static (slot-level) keys — so two sibling tasks sharing node+agent+topic can no longer cross-match. Covered by the taskId-scoped test (line 522).

**Prior P2 — missing-definition misread (Codex).** Fixed. Result now carries `definitionResolved`; when `getWorkflowForRun` returns null, `declared` is empty but flagged unavailable rather than "declares nothing." Verified `getWorkflowForRun` returns null (not throw) on a missing definition. Test at line 502.

**Prior P2 — orphan-doc framing / drop unused exports, P3 — envelope test, vocab alignment, edge-case tests.** All addressed: the orphan-doc now calls the upgrade window benign; all four result interfaces are non-exported (no premature surface — matches the internal `WorkflowSubscriptionTarget`); an end-to-end envelope round-trip test was added in node-agent-tools.test.ts; schema and MCP description both say "static"; edge-case tests added (multiple static interests, case-insensitivity, idempotent re-registration, definition-missing).

### Fresh verification (dimensions #2-#6)

- **Correctness and resilience** — the reconciliation is sound. Casing is handled consistently: the trie stores original-case topics and `subscriptionReconcileKey` is the single normalization point on both sides, so declared-to-active and persisted-to-active match regardless of case. Static keys (no taskId) and dynamic keys (with taskId) are structurally disjoint, so the two kinds never collide; the `?? 'dynamic'` default on a target's `subscriptionKind` is belt-and-braces only (both insert paths set it explicitly). `nodeFilter` is applied symmetrically across all three layers, so scoped calls stay consistent. `resolveNodeAgents` does throw on a malformed node, so the try/catch is load-bearing, not dead. `listSubscriptions`, `TopicTrie.values()`, and `listByRun` are provably read-only (no register/unregister/upsert/delete/insert). Infra throws from the runtime are caught by `onListSubscriptions`'s try/catch. No issue.
- **Impact and compatibility** — `listByRun` re-add mirrors `listBySpace` exactly (same SQL shape, parameterized, `rowToSubscription`, `ORDER BY created_at`); `TopicTrie.values()` is purely additive; conditional MCP registration matches the `list_deliveries`/`externalEventStore` sibling pattern. No contract broken.
- **Security** — `spaceId` is the trusted session/task value (captured in the `buildNodeAgentMcpServerForSession` closure, never an agent arg); the cross-space guard rejects a foreign-space `workflowRunId`; SQL is parameterized; results are run-scoped; path is read-only. Confirmed.
- **Tests and performance** — reconciliation coverage is thorough (all three layers, every mismatch direction, the P1 suppression and its orphan inverse, `topicFrom` inert exclusion, nodeId filter, cross-space guard, unknown run, taskId scoping, case-insensitivity) with meaningful assertions (specific topics, exact mismatch counts). `TopicTrie.values()` is O(total subscriptions in space) but is reached only from the manual agent-facing MCP tool path — never the tick loop, delivery, or rehydrate — so acceptable at current scale, and the inline comment flags the escape hatch. No leak/N+1. Per-test in-memory DB with afterEach close keeps isolation clean.
- **Craft and architecture** — layering (tool to callback to facade to runtime to repo/trie) matches `list_deliveries`; no dead code or unused imports; naming consistent; comments match behavior.

### P3 nits (non-blocking — optional polish)

1. **space-runtime.ts:1450 / test coverage** — the `staticMaterializable` gate treats only done/archived/cancelled as terminal, so a canonical task in review/approved/blocked keeps its statics (intentional, per the line 1205 comment about post-approval needs). That invariant is not directly tested — only `done` is. A review/approved-status test would lock in that the two gates (`registerRunInterestsFromWorkflow:1208` and `listSubscriptions:1450`) agree on the same status set, guarding against a future unilateral change. Same for the zero-task run (`pickCanonicalTaskForRun` returns null, `staticMaterializable` false).
2. **space-runtime.ts:1440** — `getWorkflowForRun(run) ?? null` is redundant (the method already returns null, not undefined), but it's a harmless defensive idiom consistent with other call sites in this file. No change needed.

No P0, P1, or P2 findings. Approving.

*(Own-PR: GitHub rejects APPROVE from the author, so this is posted as a COMMENT review carrying `Recommendation: APPROVE`.)*

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 5]
```diff
@@ -1304,6 +1402,193 @@ export class SpaceRuntime {
...
+    for (const entry of active) {
+      // A static entry is "backed" only while the canonical task is non-terminal:
+      // on a done/archived/cancelled task the lifecycle clears static interests,
+      // so any surviving static entry is stale cleanup (counted as orphan) rather
+      // than a legitimate declared interest. Without this, a terminal task with a
+      // leftover static entry reads zero across every mismatch count.
+      const backed =
+        entry.subscriptionKind === 'static'
+          ? staticMaterializable &&
+            declaredKeys.has(subscriptionReconcileKey(entry.nodeId, entry.agentName, entry.topic))
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts]: Fixed in `b38785042`: when `definitionResolved` is false a static entry's backing is unverifiable (the declaration layer is unavailable, not absent), so it is now classified `source:'unknown'` and excluded from `orphanActive` instead of a false orphan drift signal. Added `'unknown'` to the source union. Test added (`a static entry is "unknown" (not orphan) when the definition cannot be loaded`).

File: packages/daemon/src/lib/space/runtime/space-runtime.ts [hunk 6]
```diff
@@ -1304,6 +1402,193 @@ export class SpaceRuntime {
...
+      }));
+
+    // Reconcile durable ↔ trie via normalized keys. Static keys are task-
+    // independent (declared interests are slot-level); dynamic keys include the
+    // taskId so two tasks sharing a slot+topic can't cross-match.
+    const activeStaticKeys = new Set<string>();
+    const activeDynamicKeys = new Set<string>();
+    for (const entry of active) {
+      if (entry.subscriptionKind === 'static') {
+        activeStaticKeys.add(subscriptionReconcileKey(entry.nodeId, entry.agentName, entry.topic));
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/space/runtime/space-runtime.ts]: Fixed in `b38785042`: static reconciliation now compares against the canonical task. A static entry only backs a declaration when it belongs to the canonical, non-terminal task (`canonicalTaskId`); a stale entry from a duplicate/superseded task is reclassified `source:'orphan'` and counted, and no longer marks the slot's declaration active. `declared.active` now reflects only effectively-active (resolved + canonical + materializable) static entries, consistent with the per-entry source. Test added (`a static entry on a non-canonical task is orphan, not declared`).

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude (Anthropic)

> **Model:** Claude | **Client:** HyperNeo | **Provider:** Anthropic

**Verdict: APPROVE** — read-only diagnostic tool, no blocking findings.

Recommendation: APPROVE

Fresh review of head `b3878504` (own-PR, so posting as COMMENT per the workflow's own-PR convention; the marker line above carries the verdict).

### Goal & ask
Task #908 asked for a read-only MCP tool (`list_subscriptions`) that answers "is this node actually subscribed to these events?" from **durable** state, surfacing declared-vs-active drift — the single question the #896 investigation needed but couldn't get without raw SQL. Acceptance: three layers visible (declared/persisted/active), a mismatch is visible, read-only, unit tests. The diff delivers exactly this with no scope creep and no duplication of `list_deliveries` or the unrelated long-horizon `listSubscriptions`.

### Correctness & resilience — sound
Traced every branch of the reconciliation in `SpaceRuntime.listSubscriptions` (space-runtime.ts ~1408–1612):
- **Reconcile key split is correct.** Static keys omit `taskId` (declaration is slot-level); dynamic keys include it. This precisely mirrors how `registerRunInterests` (1149) stores static targets and `registerSubscription` (1224) persists dynamic rows. The taskId-scoping prevents a sibling task's trie entry from masking persisted/active drift — covered by the test at line 584.
- **`staticMaterializable` mirrors the source of truth.** The terminal-status check (`!== done/archived/cancelled`, ~1454) is identical to `registerRunInterestsFromWorkflow` (1211), so `declaredNotActive` is correctly suppressed for terminal runs where the lifecycle intentionally cleared static interests.
- **Three-way classification is honest.** `unknown` (definition can't load → backing unverifiable, not drift), `orphan` (in-trie with no durable backing), `declared`/`persisted` (backed). The canonical-task ownership check (`!canonicalTaskId || entry.taskId === canonicalTaskId`) correctly reclassifies a stale duplicate-task static entry as orphan (test line 465) rather than letting it mask drift.
- **Edge cases handled:** empty trie, null-topic (`topicFrom`) interests (inert, excluded from counts), malformed nodes (`resolveNodeAgents` → `catch { continue }`), unknown run, cross-space guard.
- **Concurrency is fine:** the method is fully synchronous — no `await` — so the trie+DB+task snapshot is atomic within one event loop tick. No torn read.
- `topicTrie.values()` walks both `exactChildren` and `globChildren`, collecting `node.values` at every node — visits the full trie, can't miss values, and dedup is unnecessary because insertion is remove-then-insert.

### Security — sound
- **Authz:** `spaceId` comes from the authenticated parent task's closure (`parentTask.spaceId`), never from agent-controlled args, and the runtime re-validates `run.spaceId !== spaceId`. Same scoping as `list_deliveries`.
- **Read-only verified:** `listSubscriptions` performs only `getRun`/`getWorkflowForRun`/`listByWorkflowRun`/`listByRun` reads plus a trie walk — zero calls to `insert`/`remove`/`upsert`/`delete*`/`prepare(...).run`. `values()` mutates nothing.
- **No injection:** `workflowRunId`/`nodeId` are agent-controlled but used only as bound DB params and in-memory equality filters, never interpolated. `listByRun` SQL is parameterized (the in-string tabs are harmless SQLite whitespace and match the file's convention).
- **No secret disclosure:** subscription rows/targets carry only routing fields (slot, topic, kind). `LongHorizonSubscriptionTarget` entries are filtered out by `isWorkflowSubscriptionTarget`.
- Same-space cross-node/run visibility is consistent with the existing trust model (cooperative peers within a Space, same as `list_deliveries`).

### Tests & performance — strong
- **17 runtime tests + 5 handler tests**, with *exact* assertions on mismatch counts, per-entry `source`/`active` classifications, and the full envelope round-trip — not smoke tests. Every subtle branch has a dedicated case: terminal-task suppression, orphan-on-terminal-leftover, unknown-when-definition-unavailable, non-canonical-task orphan, case-insensitive reconcile, taskId-scoped dynamic reconcile, topicFrom exclusion, idempotent re-registration, definition-not-resolving, #896 regression.
- **Performance is linear:** `O(declared + persisted + active)` reconcile using `Set`-based key lookups (no `Array.includes` in a loop → no accidental quadratic). `values()` is `O(total trie entries in the space)` — acceptable for an on-demand diagnostic, and the comment correctly flags run-indexing as the future path if it's ever called in a loop. `listByRun` is a single indexed single-run `SELECT`. No leaks (read-only, no handles).

### Craft & architecture — clean
- Delegation chain (handler → `onListSubscriptions` → `SpaceRuntimeService` → `SpaceRuntime` → repo) is identical in shape to `subscribe_external_event`/`list_deliveries`. Conditional MCP registration matches the sibling-tool idiom.
- The `'orphan' as const` seed (then overwritten in the reconcile loop) is a sound defensive default — any entry that escapes the loop surfaces as drift rather than appearing healthy.
- Interfaces well-named; `subscriptionReconcileKey` well-factored; JSDoc accurately describes terminal-task suppression and the `unknown` classification.

### Optional, non-blocking (P3)
One untested branch: the `try { resolveNodeAgents(node) } catch { continue }` defensive path for a malformed node (space-runtime.ts ~1466). A future test with one valid + one malformed node would confirm the tool returns a valid result excluding that node's slots (the consequence of it misfiring is only one node absent from `declared`, surfacing as `orphan`). Not worth blocking on.

No P0, P1, or P2 findings across correctness, security, tests, performance, and craft. Approving.
